client/my-sites/email/email-forwarding/email-forwarding-add-new-compact-list.jsx

```diff
@@ -0,0 +1,171 @@
+import { Button } from '@automattic/components';
+import { localize } from 'i18n-calypso';
+import PropTypes from 'prop-types';
+import { Component, Fragment } from 'react';
+import { connect } from 'react-redux';
+import { validateAllFields } from 'calypso/lib/domains/email-forwarding';
+import EmailForwardingAddNewCompact from 'calypso/my-sites/email/email-forwarding/email-forwarding-add-new-compact';
+import { composeAnalytics, withAnalytics } from 'calypso/state/analytics/actions';
+import { addEmailForward } from 'calypso/state/email-forwarding/actions';
+import getEmailForwardingLimit from 'calypso/state/selectors/get-email-forwarding-limit';
+import {
+	addEmailForwardSuccess,
+	getEmailForwards,
+	isAddingEmailForward,
+} from 'calypso/state/selectors/get-email-forwards';
+import { isRequestingSiteDomains } from 'calypso/state/sites/domains/selectors';
+import { getSiteSlug } from 'calypso/state/sites/selectors';
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+
+class EmailForwardingAddNewCompactList extends Component {
+	static propTypes = {
+		emailForwards: PropTypes.array,
+		emailForwardingLimit: PropTypes.number,
+		onConfirmEmailForwarding: PropTypes.func.isRequired,
+		onAddEmailForwardSuccess: PropTypes.func,
+		selectedDomainName: PropTypes.string.isRequired,
+	};
+
+	constructor( props ) {
+		super( props );
+		this.state = {
+			emailForwards: [ { destination: '', mailbox: '', isValid: false } ],
+			isRedirecting: false,
+			newEmailForwardAdded: false,
+		};
+	}
+
+	hasValidEmailForwards() {
+		return ! this.state.emailForwards?.some( ( forward ) => ! forward.isValid );
+	}
+
+	addNewEmailForwardWithAnalytics = ( domainName, mailbox, destination ) => {
+		withAnalytics(
+			composeAnalytics(
+				this.props.onConfirmEmailForwarding(),
+				this.props.addEmailForward( domainName, mailbox, destination )
+			)
+		);
+	};
+
+	addNewEmailForwardsClick = ( event ) => {
+		const { isSubmittingEmailForward, selectedSiteSlug, selectedDomainName } = this.props;
+
+		event.preventDefault();
+
+		if ( isSubmittingEmailForward ) {
+			return;
+		}
+
+		this.state.emailForwards?.map( ( forward ) => {
+			const { mailbox, destination } = forward;
+
+			this.addNewEmailForwardWithAnalytics(
+				selectedDomainName,
+				mailbox,
+				destination,
+				selectedSiteSlug
+			);
+		} );
+		this.setState( { newEmailForwardAdded: true } );
+	};
+
+	onAddNewEmailForward = () => {
+		this.setState( {
+			emailForwards: [ ...this.state.emailForwards, { destination: '', mailbox: '' } ],
```

Comment:
Updating the state of a component based on the current value of `this.state` may potentially lead to inconsistent component state.

This is also documented in the [React docs](https://reactjs.org/docs/react-component.html#setstate):

> `setState()` does not always immediately update the component. It may batch or defer the update until later. This makes reading `this.state` right after calling `setState()` a potential pitfall. Instead, use `componentDidUpdate` or a `setState` callback (`setState(updater, callback)`)

The easiest way to fix this is to just change `setState()` to the callback form, where it will always properly access the current state value.